aten/src/ATen/native_parse.py

```diff
@@ -98,48 +132,84 @@ def parse_arguments(args, func_decl, func_name, func_return):
-            # This enables Tensor? x=None and translates to legacy
-            # "Tensor? x={}". See [temp translations].
-            if t == 'Tensor?' and ns[1] == 'None':
-                ns[1] = "[]"  # Will translate to {} via parse_default
-            # This enables "Generator? x = None and translates to legacy
-            # "Generator* x = nullptr". See [temp translations].
-            if t == 'Generator*' and ns[1] == 'None':
-                ns[1] = 'nullptr'
-            name, default = ns[0], parse_default(ns[1])
-
-        typ = sanitize_types(t)
-        assert len(typ) == 1
-        argument_dict = {'type': typ[0].rstrip('?'), 'name': name, 'is_nullable': typ[0].endswith('?')}
-        # Enables int[x] by translating to legacy IntList[x]. See [temp translations]
-        match = re.match(r'int\[(\d+)\]', argument_dict['type'])
-        if match:
-            argument_dict['type'] = 'IntList'
-            argument_dict['size'] = int(match.group(1))
-        argument_dict['type'] = temp_type_translations(argument_dict['type'])
+        t, name, default, size, annotation = type_argument_translations(arg)
+
+        argument_dict = {'type': t.rstrip('?'), 'name': name, 'is_nullable': t.endswith('?'), 'annotation': annotation}
+        if size:
+            argument_dict['size'] = size
         if default is not None:
             argument_dict['default'] = default
-        # TODO: convention is that the ith-argument correspond to the i-th return, but it would
-        # be better if we just named everything and matched by name.
-        if is_out_fn and arg_idx < len(func_return):
-            argument_dict['output'] = True
         if kwarg_only:
             argument_dict['kwarg_only'] = True
-
         arguments.append(argument_dict)
+
+    is_out_fn = False
+    arguments_out = []
+    arguments_other = []
+    for argument in arguments:
+        if argument['type'] == "Tensor" and \
+                argument['annotation'] and \
+                re.match(r'^(.*!)$', argument['annotation']) and \
+                argument.get('kwarg_only'):
+            argument['output'] = True
+            argument['kwarg_only'] = False
+            arguments_out.append(argument)
+            is_out_fn = True
+        else:
+            arguments_other.append(argument)
+
+    arguments = arguments_out + arguments_other
+
+    name = declaration['name']
+    if is_out_fn:
+        declaration['name'] += "_out"
+
+    # Sanity checks
+
+    # TODO: convention is that the ith-argument correspond to the i-th return, but it would
+    # be better if we just named everything and matched by name.
+    for arg_idx, argument in enumerate(arguments_out):
+        assert argument['annotation'] == func_return[arg_idx]['annotation'], \
+            "For func {} writeable keyword Tensor arguments need to have a matching return Tensor. Further, " \
+            "the ith-argument needs to correspond to the i-th return.".format(name)
```

Comment:
Isn't this violated by `_copy_same_type_`? I'm curious why this isn't triggering. https://github.com/pytorch/pytorch/blob/1c0504b889d256391a5a51a3ca8dd07f41fd649d/aten/src/ATen/native/native_functions.yaml#L559

Reply:
It's not a keyword Tensor argument, but positional.